core/src/test/scala/org/apache/spark/ui/memory/MemoryListenerSuite.scala

```diff
@@ -0,0 +1,258 @@
+    val listener = new MemoryListener
+    val execId1 = "exec-1"
+    val execId2 = "exec-2"
+    val execId3 = "exec-3"
+
+    // two executors added first
+    listener.onExecutorAdded(
+      SparkListenerExecutorAdded(0L, execId1, new ExecutorInfo("host1", 1, Map.empty)))
+    listener.onExecutorAdded(
+      SparkListenerExecutorAdded(0L, execId2, new ExecutorInfo("host2", 1, Map.empty)))
+
+    // three executors running in one stage and one executor is removed before stage complete
+    listener.onStageSubmitted(MemoryListenerSuite.createStageStartEvent(1))
+    val exec1Metrics = MemoryListenerSuite.createExecutorMetrics("host-1", 1446336000L, 20, 10)
+    listener.onExecutorMetricsUpdate(MemoryListenerSuite.createExecutorMetricsUpdateEvent(
+      execId1, exec1Metrics))
+    val exec2Metrics = MemoryListenerSuite.createExecutorMetrics("host-2", 1446337000L, 15, 5)
+    listener.onExecutorMetricsUpdate(MemoryListenerSuite.createExecutorMetricsUpdateEvent(
+      execId2, exec2Metrics))
+    // one more executor added during the stage is running
+    listener.onExecutorAdded(
+      SparkListenerExecutorAdded(0L, execId3, new ExecutorInfo("host3", 1, Map.empty)))
+    val exec3Metrics = MemoryListenerSuite.createExecutorMetrics("host-3", 1446338000L, 30, 15)
+    listener.onExecutorMetricsUpdate(MemoryListenerSuite.createExecutorMetricsUpdateEvent(
+      execId3, exec3Metrics))
+    // executor 2 removed before stage complete
+    listener.onExecutorRemoved(SparkListenerExecutorRemoved(0L, execId2, ""))
+    listener.onStageCompleted(MemoryListenerSuite.createStageEndEvent(1))
+
+    listener.onExecutorRemoved(SparkListenerExecutorRemoved(0L, execId1, ""))
+    listener.onExecutorRemoved(SparkListenerExecutorRemoved(0L, execId3, ""))
+
+    // the completedStagesToMem will maintain the metrics of both the removed executors and new
+    // added executors
+    val mapForStage1 = listener.completedStagesToMem.get((1, 0)).get
+    assert(mapForStage1.size === 3)
+    val memInfo1 = mapForStage1.get(execId1).get
+    val memInfo2 = mapForStage1.get(execId2).get
+    val memInfo3 = mapForStage1.get(execId3).get
+    val transMetrics1 = memInfo1.transportInfo.get
+    val transMetrics2 = memInfo2.transportInfo.get
+    val transMetrics3 = memInfo3.transportInfo.get
+    assert((20, 10, MemTime(20, 1446336000), MemTime(10, 1446336000)) === (
+      transMetrics1.onHeapSize,
+      transMetrics1.offHeapSize,
+      transMetrics1.peakOnHeapSizeTime,
+      transMetrics1.peakOffHeapSizeTime))
+    assert((15, 5, MemTime(15, 1446337000), MemTime(5, 1446337000)) === (
+      transMetrics2.onHeapSize,
+      transMetrics2.offHeapSize,
+      transMetrics2.peakOnHeapSizeTime,
+      transMetrics2.peakOffHeapSizeTime))
+    assert((30, 15, MemTime(30, 1446338000), MemTime(15, 1446338000)) === (
+      transMetrics3.onHeapSize,
+      transMetrics3.offHeapSize,
+      transMetrics3.peakOnHeapSizeTime,
+      transMetrics3.peakOffHeapSizeTime))
+  }
```

Comment:
slightly higher-level question -- is anything gained by breaking this into the different test cases?  I feel like you could just take this final test case, add some more stages at the end with no metric updates, and also add some asserts on `listener.activeStagesToMem` etc., and that would cover all the other test cases.

I'm all for having more complete coverage with more test cases, but not sure we gain anything here.  what do you think?